Comparison of revision branch "preter_review" to master

BristolArea8 opened the pull request:
The preter_review branch is now updated with a working version of the add_checkin function in views.py.  We now have a new working version of the #SAVE THE LEGS# section of the add_checkin function in views.py.  In order to get all test scenarios working, I needed to start from scratch on that section.  Where previously we had 4 separate if/else statements in there, we now have one large nested if.  (the "one big if to rule them all" approach).  The code is not as efficiently-constructed as the prior approach (there are now more lines of code), but it works, and its easy to follow the logic for each test case.

This new version of the simplified checkin form:
    - correctly renders the radio buttons and reveals/hides data entry fields based on the state of the radio button(s), and
    - correctly gathers and stores the data for all test scenarios involving all possible combinations of data entry, and
    - also doesn't incorrectly raise an incomplete form alert when the the form is in fact complete (which had been happening after our last Monday night session).

But, when the form is not complete, it doesn't raise a form incomplete error.  I.e., one can submit the form without having selected a mode from the dropdown list.  Something isn't working correctly with the is_valid form function, for the mode field.  I think mode is defined as required, so possibly the default is not actually=NONE when user doesn't select a value from the mode?... so the is_valid function thinks the mode has been filled in?

Commit:
simplify checkin process - a work in progress

Commit:
refactor leg saving and add more comments.

Commit:
revised save the legs section of add_checkin function, to ensure if statements working for all test cases

pwnage101 commented:
I'm still working though the first couple of blocks, but so far the logic looks good to me.  Only ask so far would be to amend df79213109fe43fac6662c073d73d7e86056df41 to replace the tabs with spaces.  This is not just for readability---python rules are strict about indentation.

pwnage101 commented:
Also, can you add comments to each of the 8 conditional blocks created by:

```
if request.POST['normal_same_as_reverse'] == 'True':
    if request.POST['normal_same_as_walkride'] == 'True':
        if request.POST['walkride_same_as_reverse'] == 'True':
```

in order to indicate the status of these three POST variables?  Something like this would be sufficient:

```
# normal_same_as_reverse = True/False
# normal_same_as_walkride = True/False
# walkride_same_as_reverse = True/False
```

pwnage101 commented:
per our in-person conversation: to update github pull requests, amend your commit and force push:
1. edit code in your preter_review branch
2. `git add ...`
3. `git commit --amend`
4. `git push --force`

pwnage101 commented on survey/views.py:
```diff
@@ -75,69 +75,164 @@ def add_checkin(request):
+				    # need to correct for the hidden fields
+                        for form in leg_formset_NormalFW:
+						    leg = form.save(commit=False)
+						    leg.day = 'n'
```
delete this line (149)

pwnage101 commented on survey/forms.py:
```diff
@@ -59,11 +59,11 @@ def __init__(self, *args, **kwargs):
-            "Or, if you are not "
-            "employed, other destination")
+        	"Or, if you are not "
+        	"employed, other destination")
```
don't modify whitespace here.

pwnage101 commented:
Per our in-person conversaion, how to modify commits in this pull request which are not the latest commit:
1. `git pull` (double check you have the latest version of the remote branch)
2. `git rebase --interactive HEAD^^^` (use more or less carets depending on which commit to change)
3. change "pick" to "edit" on the commit you want to change
4. save and quit
5. make changes you intend on making and add to index
   - if the changes are to undo something, that can be easily done using `git checkout --patch HEAD^`
6. `git commit --amend`
7. `git rebase --continue`
8. `git status` (double check this says your branch has diverged from the remote)
9. `git push --force`

pwnage101 commented on survey/views.py:
```diff
@@ -75,69 +75,164 @@ def add_checkin(request):
+                            request.POST, instance=commutersurvey, prefix='ntw')
+                    # set normal legs from work = w/r legs to work
+                        leg_formset_NormalFW = MakeLegs_NormalTW(
+						    request.POST, instance=commutersurvey, prefix='ntw')
```
comment doesn't match code here (143-144)

pwnage101 commented on survey/views.py:
```diff
@@ -234,6 +232,9 @@ def add_checkin(request):
 
+        normal_copy = NormalFromWorkSameAsAboveForm()
+        wrday_copy = WalkRideFromWorkSameAsAboveForm()
+        commute_copy = NormalIdenticalToWalkrideForm()
```
Cristen, are these three lines necessary (235-237)?  If not, delete them!

Commit:
tighten radio button logic! set to required=False (or else the form wants them all to be "YES") and clean up the JS toggling code

Commit:
re-introduce RequiredFormset and clean up some non-leg-related view logic

pwnage101 commented:
Thanks!  In my testing, a valid main form but invalid leg causes an uncaught ValueError:

ValueError at /checkin/
The Leg could not be created because the data didn't validate.

This is a good sign, but still needs to be caught.  I haven't really looked into it yet, so I don't have a solution.

dougielong commented:
You need to validate the form before trying to save it.

`leg_formset_WRFW = MakeLegs_WRTW(
    request.POST, instance=commutersurvey, prefix='wtw')

for form in leg_formset_WRFW:

```
### Validate Form before trying to save ######################
if form.is_valid():
    leg = form.save(commit=False)
    leg.direction = 'fw'
else:
    #raise ValidationError('Invalid value', code='invalid')
    pass
###############################################################`
```

pwnage101 commented:
assigned to self, will try to throw in some validation checks (per IdeaHackDouglas comment above)

pwnage101 committed:
fix bad indentation

Sorry, but this indentation is a recipe for disaster.  It needs to be fixed
before I can read the code.

pwnage101 committed:
validate legs immediately after user leg data collection

Validation needs to happen earlier than immediately before the step which
renders the thank-you page.  It needs to happen immediately after collecting
new user leg data, since some leg formsets are re-used when copying into other
leg formsets which requires that they already be valid.

pwnage101 commented:
I believe the validation problems are now solved.  I am requesting code review and somebody to test.

thecristen commented:
From what I can tell it works great upon successful submission on first attempt.
However...

Test case: 
Walk/Ride Day from work: Driving alone 11 minutes
(W/R to work same as from work = No)
Walk/Ride Day to work: Driving alone 9 minutes + Driving alone 14 minutes
(Normal same as W/R = No)
Normal from work: _leaving fields empty_
(Normal to work same as from work = Yes)

Takes me back to the form with errors in the right place. However,
- W/R to work same as from work = reset to Yes instead of user-selected No _and associated leg forms are showing instead of hidden_ 
- When Normal from work is corrected and the form re-submitted, the result entered in the DB is Driving alone 11 minutes copied to all 4 parts of the journey!

Commit:
fix janky handling of hiding/showing sections of checkin form

Commit:
consolidate LegForm classes finally

Commit:
comments, another validation check, adjust nesting for final if statement

thecristen commented:
:rotating_light: I think I fixed my failing test case!

It might very well be without validation problems now. Requesting some inspection and more testing!

pwnage101 commented:
Yay! I will test today.

pwnage101 commented:
There's an issue with this test case:
- W/R to work: filled in
- W/R from work same as to work: Yes
- Normal same as W/R: No
- Normal from work: filled in
- Normal to work same as from work: empty

Now click submit and it should throw you back into the same page with a red error in the field you left empty (so far, so good).  Fill it in and click submit.  I'm getting this traceback:

```
Traceback:
File "/home/sankey/.virtualenvs/greenstreets/local/lib/python2.7/site-packages/django/core/handlers/base.py" in get_response
  111.                     response = wrapped_callback(request, *callback_args, **callback_kwargs)
File "/shared/greenstreets/survey/views.py" in add_checkin
  54.             request.POST, instance=Commutersurvey(), prefix='wfw')
File "/shared/greenstreets/survey/forms.py" in __init__
  119.         for form in self.forms:
File "/home/sankey/.virtualenvs/greenstreets/local/lib/python2.7/site-packages/django/utils/functional.py" in __get__
  55.         res = instance.__dict__[self.func.__name__] = self.func(instance)
File "/home/sankey/.virtualenvs/greenstreets/local/lib/python2.7/site-packages/django/forms/formsets.py" in forms
  141.         forms = [self._construct_form(i) for i in xrange(self.total_form_count())]
File "/home/sankey/.virtualenvs/greenstreets/local/lib/python2.7/site-packages/django/forms/formsets.py" in total_form_count
  114.             return min(self.management_form.cleaned_data[TOTAL_FORM_COUNT], self.absolute_max)
File "/home/sankey/.virtualenvs/greenstreets/local/lib/python2.7/site-packages/django/forms/formsets.py" in management_form
  96.                     code='missing_management_form',

Exception Type: ValidationError at /checkin/
Exception Value: [u'ManagementForm data is missing or has been tampered with']
```

Also, I noticed this in the POST data, but I'm not sure if it is relevant (could be a side-effect of the problem):

```
wtw-0-mode: u''
```

thecristen commented:
> Normal to work same as from work: empty

Does this mean you left the radio button empty? Or selected No and left the associated legs empty?

pwnage101 commented:
bleh, that's a typo in the comment.  Here's the test case again:
- W/R to work: filled in
- W/R from work same as to work: Yes
- Normal same as W/R: No
- Normal to work: filled in
- Normal from work same as to work: No
- Normal from work: empty

click submit, fill in Normal-from-work, then click submit again.

edit: i just typoed this comment again, and edited again.  Hope you see this latest edit.

thecristen commented:
Thanks for the update. Let's keep investigating.
I wonder if today might be a good time to simplify some of this code (make a few reusable functions and whatnot).

thecristen commented:
I've been trying to fix this all afternoon and continue t have problems. I suspect our way of copying one formset to another is causing some issue. 

I had a radical thought - what if we stop this server side formset copying mess, and handle all the leg copy client side with JavaScript? We could attach it to change event handlers for the radio buttons to have it copy the form values from one element to another (I wouldn't recommend replacing the elements outright).

Commit:
handle all form copying with JS instead of Django

Commit:
use cookies for Employer and Team fields

Commit:
adjust carbon savings for clarity

thecristen commented:
Update: I followed through on the thought I expressed in my last comment. Fixes any failing test cases previously noted here.
Since we're in a bit of a time crunch I'll probably merge this in a few hours if all is well.

pwnage101 commented:
I will start testing now and then I may have a dinner to attend in about 30 minutes.  Will report back before i leave.

pwnage101 commented:
Seems to pass anything I can come up with.  I also reviewed all the code changes since my commit, and it LGTM.
